internal/orchestrator/subscriptions.go

```diff
@@ -134,3 +134,61 @@ func (or *orchestrator) GetSubscriptionByIDWithStatus(ctx context.Context, id st
 
 	return subWithStatus, nil
 }
+
+func (or *orchestrator) GetSubscriptionEventsHistorical(ctx context.Context, subscription *core.Subscription, filter ffapi.AndFilter) ([]*core.EnrichedEvent, *ffapi.FilterResult, error) {
+
+	// Internally we need to know the limit/count from the inbound filter
+	inboundFilterOptions, err := filter.Finalize()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var ssFilters = &ffapi.QueryFields{}
+	fb := ssFilters.NewFilter(ctx)
+	ssFilter := fb.And()
+
+	// This references the final offset on the FILTERED list
+	finalDesiredOffset := inboundFilterOptions.Skip + inboundFilterOptions.Limit
+	var subscriptionFilteredEvents []*core.EnrichedEvent
+
+	internalLimit := 200
+	internalSkip := 0
+	ssFilter.Limit(uint64(internalLimit))
+
+	for len(subscriptionFilteredEvents) < int(finalDesiredOffset) {
```

Comment:
I think the easiest solution to the `o(n)` safety issue (which is a significant problem) is to have a server-configurable (not API configurable) limit on the max scan length for this.

I would suggest something like 1000 by default.

If the query reaches this limit before either:
1. Reaching the end of the table
2. Reaching the full `limit` specified

Then it returns with an error result. For example:

- `Event scan limit reached after skipping X events. Please restrict your query to a narrower range`

It's far from a perfect solution, but in the vast majority of cases I think it gives a solution.
Either:
- You have lots of events that match your subscription, so your page size will find 25 (or whatever) events in the last 1000 events
- You are hunting for one specific event, so the pre-filter (discussed in my comment above) will sub-select to less than 1000 records. Either with a time-range, or with a spear-fishing approach (like looking for one specific event)

Reply:
The implementation for these changes is split across commits https://github.com/hyperledger/firefly/pull/1452/commits/de6d46b3ef8cbb763ed4d5668fda220ed693527e and https://github.com/hyperledger/firefly/pull/1452/commits/7e22a2cc20ebb54d4c0bfa00f406b9343dd1d116.

This approach means that:
 - Supplying a skip/limit in the initial query now applies specifically to the amount of unfiltered events that are processed on the backend and you cannot control how many results you get from the API*1
 - If the amount of records being indexed exceeds the total amount configured on the namespace, no results are returned to the user, and instead they get an error indicating that they need to refine their query

---

Crucially, this approach presents a new problem worth noting: **how do I as the caller of the API, know how to adjust my skip/limit so that I can receive the next _sequentially filtered_ event?**

Suppose 1000 unfiltered events, with 2 matching events at position [0:1], and a matching event at [999]. My first query might be `skip:0 limit:100` which would return me events 0 and 1, but how would I get event 2 which is at unfiltered index 999?

Currently the answer to this is: 'Keep adjusting the skip/limit until you find it' which is fair, but depending on the use-case could be a frustrating answer. For us being able to supply information to make this easier for users is difficult because we as FireFly don't have the intelligence required to know where that index is.

---

*1 - The skip/limit allows the user to configure how many unfiltered events are indexed, and then returns all of the matching events within that bound, we then assume that the user will then limit this down to be the number of events they want to search on their side.